### ISSUE TITLE
Why don't we have `TypInfoNoBound`?

### ISSUE THREAD

**Issue by shawntabrizi** [paritytech/scale-info#210]
In the `polkadot-sdk`, we have many derive macros which can handle the generic `T: Config`.

This is done through the `*NoBound` derives:

- `CloneNoBound`
- `ParitialEqNoBound`
- `DefaultNoBound`
- etc...

Why don't we have this for `TypeInfo`, rather than having to explain and use `#[scale_info(skip_type_params(T))]`?

If there is no reason, this is a feature request, and I think will provide a better developer experience.

**Comment by bkchr** [comment]
Probably just because no one has done it yet. Not sure we want to add it here or also just to `polkadot-sdk`?

**Comment by shawntabrizi** [comment]
I guess `polkadot-sdk` if that is where the pattern is coming from, but no strong opinions

**Comment by gui1117** [comment]
I think `TypeInfo` implementation similar to `Encode`/`Decode` where the types in fields and variants are bounded instead of generics attributes.

Maybe what we want here is `TypeInfoNoTypeParam`?

(I wonder if we should have made the type paramater ignored as a default.)

I can implement `TypeInfoNoTypeParam`.

**Comment by bkchr** [comment]
We should probably stick with the name `TypeInfoNoBound` to stay consistent.

**Comment by jsdw** [comment]
I agree that it should probably live in `polkadot-sdk` if that's where the other `NoBounds` proc macros are.

IIRC this library also tries something smarter already where it skips adding attrs for type params that aren't used in the type (can't remember exactly the logic offhand), so do you need to use `skip_type_params` often? (the `dumb_trait_bounds` attribute makes it act more like most other libraries and just dumbly add bounds for all type params).

**Comment by shawntabrizi** [comment]
> so do you need to use skip_type_params often

In runtime development, all the time, because it is often you create objects generic over `T: Config`

**Comment by gui1117** [comment]
> IIRC this library also tries something smarter already where it skips adding attrs for type params that aren't used in the type (can't remember exactly the logic offhand), so do you need to use `skip_type_params` often? (the `dumb_trait_bounds` attribute makes it act more like most other libraries and just dumbly add bounds for all type params).

The type parameters are part of the info. This was made in case they contains some information, like `BoundedVec` for example we could want the information of the generic which limit the number of value.

**Comment by gui1117** [comment]
> We should probably stick with the name `TypeInfoNoBound` to stay consistent.

Do you mean:
* naming it `TypeInfoNoBound` but just skip all type parameters under the hood?
* Actually bound nothing and also skip all type parameters under the hood?

**Comment by shawntabrizi** [comment]
We really need just need of homogeneously named derive macros which work on `MyObject<T: Config>`.

So I think just naming `NoBound` but having it do as much as possible is right.

TBH it is not clear to me what happens when you "bound nothing".

**Comment by bkchr** [comment]
> * You want to use `Default` but having issues? Use `DefaultNoBound`
> 
> * You want to use `TypeInfo` but having issues? Use `TypeInfoNoBound`
> 
> * You want to use `MaxEncodedLen` but having issues? Use `MaxEncodedLenNoBound`

None of this is correct. Just to tell people if you have problems do X, will not solve their problems. These macros exists out of specific reason.

> TBH it is not clear to me what happens when you "bound nothing".

These macros should be used if none of the generic types will actually be required to implement the trait. For example for `Default`, because the generic type is not a member of the struct.

**Comment by gui1117** [comment]
Yes the current state of the art is: for a type `struct Foo<T: Config> (T::AccountId);`
* standard derive macros like `Default` will bound the generics of the type: `T: Debug` and expand its code.
* Basti-style macros: `Encode`, `Decode`, `ScaleInfo` will bound the fields (with some specifics for recursion): `T::AccountId: Encode`. and expand its code.
* derive no bound macros: will bound nothing but copy the bound as written on the type declaration so: `T: Config`, and expand its code (useful for types which are meant to be used in only one context).
* derivative crate: https://crates.io/crates/derivative : AFAIK same as standard derive macros but allows to specify bounds manually.

**Comment by gui1117** [comment]
Maybe we can call it `TypeInfoNoParam` and by default it skips all type generic parameter.

If I could go in the past I would make that the default is to skip all the generic parameter, and only add them with `#[scale_info(add_type_params)]` explicit attribute.

Most of the time those parameter gives no interesting information. Only in `BoundedVec` or such types it is useful to have them.
